cs/src/Management/TunnelManagementClient.cs

```diff
@@ -1335,5 +1325,48 @@ public async Task<NamedRateStatus[]> ListUserLimitsAsync(CancellationToken cance
         {
             return string.IsNullOrEmpty(ApiVersion) ? null : $"api-version={ApiVersion}";
         }
+
+        /// <summary>
+        /// Copy access tokens from the request object to the result object, except for any
+        /// tokens that were refreshed by the request.
+        /// </summary>
+        /// <remarks>
+        /// This intentionally does not check whether any existing tokens are expired. So
+        /// expired tokens may be preserved also, if not refreshed. This allows for better
+        /// diagnostics in that case.
+        /// </remarks>
+        private static void PreserveAccessTokens(Tunnel requestTunnel, Tunnel? resultTunnel)
+        {
+            if (requestTunnel.AccessTokens != null && resultTunnel != null)
+            {
+                resultTunnel.AccessTokens ??= new Dictionary<string, string>();
+                foreach (var scopeAndToken in requestTunnel.AccessTokens)
```

Comment:
Would it make sense to check token expiration and do not preserve expired tokens?

Reply:
See comments on line 1334.

I don't think that would help anything, and it would make diagnosing expired-token problems more difficult.